### PR TITLE
Refactor references NS to use consistent and explicit variable and method names

### PR DESCRIPTION
**Pull request by georgehristov**
Related to solving #662

### CODE REVIEW

File: src/Reference.php [hunk 1]
```diff
@@ -11,7 +11,7 @@
  *
  * It's possible to extend the basic reference with more meaningful references.
  *
- * @property Model $owner
+ * @property Model $owner definition of "our model"
```
Comment (mvorisek) [inline src/Reference.php:14]: of course, remove that ;-)

Comment (mvorisek) [inline src/Reference.php]: simply (our) model

File: src/Reference.php [hunk 2]
```diff
@@ -41,7 +41,7 @@ class Reference
     public $link;
 
     /**
-     * Definition of the destination model, that can be either an object, a
+     * Definition of the destination (their) model, that can be either an object, a
```
Comment (mvorisek) [inline src/Reference.php:44]: simply "their model"

File: src/Reference.php [hunk 3]
```diff
@@ -100,59 +100,81 @@ public function getDesiredName(): string
...
 
+    public function getOurModel(): Model
+    {
+        return $this->owner;
+    }
+
+    /**
+     * @deprecated use getTheirModel instead - will be removed in dec-2020
+     */
+    public function getModel($defaults = []): Model
```
Comment (mvorisek) [inline src/Reference.php:113]: makes sense if this was returning `getTheirModel()`

File: src/Reference.php [hunk 4]
```diff
@@ -100,59 +100,81 @@ public function getDesiredName(): string
...
-                $this->table_alias = $this->link;
-                $this->table_alias = preg_replace('/_' . ($this->owner->id_field ?: 'id') . '/', '', $this->table_alias);
-                $this->table_alias = preg_replace('/([a-zA-Z])[a-zA-Z]*[^a-zA-Z]*/', '\1', $this->table_alias);
-                if (isset($this->owner->table_alias)) {
-                    $this->table_alias = $this->owner->table_alias . '_' . $this->table_alias;
-                }
-            }
-            $defaults['table_alias'] = $this->table_alias;
-        }
+        $defaults['table_alias'] = $defaults['table_alias'] ?? $this->getTableAlias();
```
Comment (mvorisek) [inline src/Reference.php]: if === null, then assing with method

do not assign the same value to vars, this is really a bad practise hard to read

File: src/Reference.php [hunk 5]
```diff
@@ -100,59 +100,81 @@ public function getDesiredName(): string
...
 
         // if model is Closure, then call it and return model
         if (is_object($this->model) && $this->model instanceof \Closure) {
-            $c = ($this->model)($this->owner, $this, $defaults);
+            $closure = ($this->model)($this->getOurModel(), $this, $defaults);
```
Comment (mvorisek) [inline src/Reference.php]: this does not return probably another closure, so wrong var name!

File: src/Reference.php [hunk 6]
```diff
@@ -100,59 +100,81 @@ public function getDesiredName(): string
...
+            return $this->addToPersistence($closure, $defaults);
         }
 
         // if model is set, then return clone of this model
         if (is_object($this->model)) {
-            $c = clone $this->model;
+            $theirModel = clone $this->model;
 
-            return $this->addToPersistence($c, $defaults);
+            return $this->addToPersistence($theirModel, $defaults);
```
Comment (mvorisek) [inline src/Reference.php:140]: probably clone $this->getTheirModel() here

File: src/Reference.php [hunk 7]
```diff
@@ -100,59 +100,81 @@ public function getDesiredName(): string
...
+            $defaults = array_merge($modelDefaults, $defaults);
         } elseif (is_string($this->model)) {
-            $model = [$this->model];
+            $theirModel = [$this->model];
         } else {
-            $model = $this->model;
+            $theirModel = $this->model;
+        }
+
+        if (!$theirModel instanceof Model) {
```
Comment (mvorisek) [inline src/Reference.php]: drop this if, always true

File: src/Reference.php [hunk 8]
```diff
@@ -100,59 +100,81 @@ public function getDesiredName(): string
...
+            $theirModel = $this->factory($theirModel, $defaults);
         }
 
-        if (!$model instanceof Model) {
-            $model = $this->factory($model, $defaults);
+        return $this->addToPersistence($theirModel, $defaults);
+    }
+
+    protected function getTableAlias(): string
```
Comment (mvorisek) [inline src/Reference.php:164]: make it private and name it better - getTableAlias can not return different result than `table_alias`!

File: src/Reference.php (hunk 4 above)
Comment (georgehristov) [reply src/Reference.php]: In new php versions this can be updated to:
$var ??= 'value';

File: src/Reference.php (hunk 3 above)
Comment (georgehristov) [reply src/Reference.php:113]: Yes, it was

File: src/Reference.php [hunk 9]
```diff
@@ -100,59 +100,81 @@ public function getDesiredName(): string
...
+    {
+        if (!$this->table_alias) {
```
Comment (mvorisek) [inline src/Reference.php:164]: seems like it should be moved to init (and then drop this function)

File: src/Reference.php (hunk 4 above)
Comment (georgehristov) [reply src/Reference.php]: I already know what you are going to comment beforehand when writing the code :)

File: src/Reference/ContainsOne.php [hunk 10]
```diff
@@ -82,24 +83,21 @@ public function init(): void
      */
     protected function getDefaultPersistence($model)
```
Comment (mvorisek) [inline src/Reference/ContainsOne.php:84]: can we drop this?

File: src/Reference.php (hunk 9 above)
Comment (georgehristov) [reply src/Reference.php:164]: I just introduced it. Makes code cleaner and it is called from other places as well

File: src/Reference/ContainsOne.php [hunk 11]
```diff
@@ -82,24 +83,21 @@ public function init(): void
...
-        $row = $m->get($this->our_field) ?: [];
-        //$row = $m->persistence->typecastLoadRow($m, $row); // we need this typecasting because we set persistence data directly
-
-        $data = [$this->table_alias => $row ? [1 => $row] : []];
-        $p = new ArrayOfStrings($data);
+        $row = $ourModel->get($this->our_field) ?: [];
+        //$row = $ourModel->persistence->typecastLoadRow($ourModel, $row); // we need this typecasting because we set persistence data directly
 
-        return $p;
+        return new Persistence\ArrayOfStrings([$this->getTableAlias() => $row ? [1 => $row] : []]);
```
Comment (mvorisek) [inline src/Reference/ContainsOne.php]: why not 20? Isn't this ID?

File: src/Reference/HasMany.php [hunk 12]
```diff
@@ -53,11 +54,12 @@ protected function referenceOurValue(): Field
...
     {
-        return $this->getModel($defaults)
-            ->addCondition(
-                $this->their_field ?: ($this->owner->table . '_' . ($this->owner->id_field ?: 'id')),
-                $this->getOurValue()
-            );
+        $ourModel = $this->getOurModel();
+
+        return $this->getTheirModel($defaults)->addCondition(
+            $this->their_field ?: ($ourModel->table . '_' . ($ourModel->id_field ?: 'id')),
```
Comment (mvorisek) [inline src/Reference/HasMany.php]: we should have more like two protected methods for build and parse (I think related with table alias regexes above)

File: src/Reference/HasOne.php [hunk 13]
```diff
@@ -164,8 +164,10 @@ public function init(): void
             $this->our_field = $this->link;
         }
 
-        if (!$this->owner->hasField($this->our_field)) {
-            $this->owner->addField($this->our_field, [
+        $ourModel = $this->getOurModel();
+
+        if (!$ourModel->hasField($this->our_field)) {
```
Comment (mvorisek) [inline src/Reference/HasOne.php:169]: what is this condition for?

File: src/Reference/HasOneSql.php [hunk 14]
```diff
@@ -86,31 +89,31 @@ function (Model $m) use ($their_field) {
      * addFields(['from', 'to'], ['type' => 'date']);
      *
      * @param array $fields
-     * @param array $defaults
+     * @param array $groupDefaults
      *
      * @return $this
      */
-    public function addFields($fields = [], $defaults = [])
+    public function addFields($fields = [], $groupDefaults = [])
```
Comment (mvorisek) [inline src/Reference/HasOneSql.php]: `fieldsDefaults`

File: src/Reference/HasOneSql.php [hunk 15]
```diff
@@ -86,31 +89,31 @@ function (Model $m) use ($their_field) {
...
-                        ->addMoreInfo('field', $field)
-                        ->addMoreInfo('alias', $alias);
+                        ->addMoreInfo('fieldName', $fieldName)
+                        ->addMoreInfo('alias', $defaults);
                 }
-                $alias = $alias[0];
+                $defaults = $defaults[0];
             } else {
-                $d = $defaults;
+                $field = $groupDefaults;
```
Comment (mvorisek) [inline src/Reference/HasOneSql.php]: to be removed, right? (and if cond above)

File: src/Reference.php (hunk 4 above)
Comment (mvorisek) [reply src/Reference.php]: this does not have any side effect, I will prefer if now, but I can live with it

it actually have - if table alias is not defined, so with array access, you "should always" use if and if you want set a fallback with `??` if key is not guaranteed to exist (like here, I think), then to be precise now you should use `if (isset($arr[k]))` and then assing. But keep this like it is now.

File: src/Reference.php (hunk 9 above)
Comment (mvorisek) [reply src/Reference.php:164]: no, you coded a getter method that modify the property when called

it should be once in init (and optinally you can define private method `initTableAlias(): void`)

File: src/Reference/HasOne.php (hunk 13 above)
Comment (georgehristov) [reply src/Reference/HasOne.php:169]: Only add field if not existing in model I guess

File: src/Reference/ContainsOne.php (hunk 10 above)
Comment (georgehristov) [reply src/Reference/ContainsOne.php:84]: It is used in sub-classes so I would not touch it

File: src/Reference/HasOne.php (hunk 13 above)
Comment (mvorisek) [reply src/Reference/HasOne.php:169]: i see "if not", probably right...

File: src/Reference/ContainsOne.php (hunk 10 above)
Comment (mvorisek) [reply src/Reference/ContainsOne.php:84]: I think it should be removed

Comment (georgehristov) [reply src/Reference/ContainsOne.php:84]: You can work on it if you like

File: src/Reference/HasMany.php (hunk 12 above)
Comment (georgehristov) [reply src/Reference/HasMany.php]: Also here - feel free to implement your ideas in this PR

Comment (mvorisek) [reply src/Reference/HasMany.php]: related with https://github.com/atk4/data/pull/675#discussion_r460437912 , this should be probably solved in this PR

File: src/Reference/ContainsOne.php (hunk 10 above)
Comment (mvorisek) [reply src/Reference/ContainsOne.php:84]: on Tuesday we start new icescream shop, not that much time now ;-)

I would like this to be solved - but it is not directly related to this PR, thus is has to be not solved now

File: src/Reference.php (hunk 9 above)
Comment (georgehristov) [reply src/Reference.php:164]: Done

File: src/Reference/HasOneSql.php (hunk 14 above)
Comment (georgehristov) [reply src/Reference/HasOneSql.php]: Done

File: src/Reference/HasOneSql.php (hunk 15 above)
Comment (georgehristov) [reply src/Reference/HasOneSql.php]: Simplified

File: src/Reference/HasMany.php (hunk 12 above)
Comment (georgehristov) [reply src/Reference/HasMany.php]: This PR is about pure renaming. Let's merge this PR and you look into it when you have time and willingness for it.

**Comment by georgehristov** [comment]
@mvorisek let's merge this now as pure var name refactoring.
We also still need to place this in the "correct" NS under Model\Reference but it is a matter of further PRs

**Review by DarkSide666 — APPROVED** [review]
Very good work on making code more readable. Especially their/our model. That was always a pain to not mix these two.

**Pull request merged**